lib/charms/kratos_external_idp_integrator/v1/kratos_external_provider.py

```diff
@@ -0,0 +1,604 @@
+        yield from self.root
+
+    def __getitem__(self, idx: int) -> RequirerProvider:
+        return self.root[idx]
+
+    def __len__(self) -> int:
+        return len(self.root)
+
+
+class RelationReadyEvent(EventBase):
+    """Event to notify the charm that the relation is ready."""
+
+    def snapshot(self) -> dict:
+        """Save event."""
+        return {}
+
+    def restore(self, snapshot: dict) -> None:
+        """Restore event."""
+        pass
+
+
+class RedirectURIChangedEvent(EventBase):
+    """Event to notify the charm that the redirect_uri changed."""
+
+    def __init__(self, handle: Handle, redirect_uri: str) -> None:
+        super().__init__(handle)
+        self.redirect_uri = redirect_uri
+
+    def snapshot(self) -> dict:
+        """Save redirect_uri."""
+        return {"redirect_uri": self.redirect_uri}
+
+    def restore(self, snapshot: dict) -> None:
+        """Restore redirect_uri."""
+        self.redirect_uri = snapshot["redirect_uri"]
+
+
+class ExternalIdpProviderEvents(ObjectEvents):
+    """Event descriptor for events raised by `ExternalIdpProvider`."""
+
+    ready = EventSource(RelationReadyEvent)
+    redirect_uri_changed = EventSource(RedirectURIChangedEvent)
+
+
+class ExternalIdpProvider(Object):
+    """Forward client configurations to Identity Broker."""
+
+    on = ExternalIdpProviderEvents()
+
+    def __init__(self, charm: CharmBase, relation_name: str = DEFAULT_RELATION_NAME) -> None:
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+
+        events = self._charm.on[relation_name]
+        self.framework.observe(events.relation_joined, self._on_provider_endpoint_relation_joined)
+        self.framework.observe(
+            events.relation_changed, self._on_provider_endpoint_relation_changed
+        )
+        self.framework.observe(
+            events.relation_departed, self._on_provider_endpoint_relation_departed
+        )
+
+    def _on_provider_endpoint_relation_joined(self, event: RelationJoinedEvent) -> None:
+        self.on.ready.emit()
+
+    def _on_provider_endpoint_relation_changed(self, event: RelationChangedEvent) -> None:
+        if not event.app:
+            return
+
+        relation_data = event.relation.data[event.app]
+        if not (providers := relation_data.get("providers")):
+            return
+
+        if not (data := RequirerProviders.model_validate(json.loads(providers))):
+            return
+
+        self.on.redirect_uri_changed.emit(redirect_uri=data[0].redirect_uri)
+
+    def _on_provider_endpoint_relation_departed(self, event: RelationDepartedEvent) -> None:
+        self.on.redirect_uri_changed.emit(redirect_uri="")
+
+    def is_ready(self) -> bool:
+        """Checks if the relation is ready."""
+        return self._charm.model.get_relation(self._relation_name) is not None
+
+    def create_provider(self, providers: Providers) -> None:
```

Comment:
nit(wording): this function creates multiple providers, so maybe rename it to `create_providers`?